File: cpp/leetcode/字母异位词分组/1.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> anagram_map;
        // const string& s 不修改原字符串
        for (const string& s : strs) {
            string key = s;
            // 对字符串进行排序，key="eat"->key="aet",key="tae"->key="aet",key="ate"->key="aet"
            sort(key.begin(), key.end());
            // unordered_map(哈希表)查找与自动创建
            // 如果key存在在哈希表中，则返回与之对应的"value"的引用
            // 如果key不存在在哈希表中，则创建一个新的key-value对，键为当前的key,value为vector<string>类型的空向量，并返回该空向量的引用
            anagram_map[key].push_back(s);
            //{"abt": ["bat"], "ant": ["tan", "nat"], "aet": ["eat", "tea", "ate"]}
        }
        vector<vector<string>> result;
        for (auto& pair : anagram_map) {
            result.push_back(pair.second);
            // result = [["bat"], ["tan", "nat"], ["eat", "tea", "ate"]]
        }
        return result;
    }
};

int main() {
    Solution solution;
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> result = solution.groupAnagrams(strs);

    cout << "[";
    for (size_t i = 0; i < result.size(); ++i) {
        cout << " [";
        for (size_t j = 0; j < result[i].size(); ++j) {
            cout << "\"" << result[i][j] << "\"";
            if (j < result[i].size() -1) cout << ", ";
        }
        cout << "]";
        if (i < result.size() - 1) cout << ",";
    }
    cout << "]" << endl;

    return 0;
}

File: cpp/leetcode/字母异位词分组/1.py
import collections
from typing import List
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_map = collections.defaultdict(list)

        for s in strs:
            key = "".join(sorted(s))

            anagram_map[key].append(s)

        return list(anagram_map.values())

if __name__ == "__main__":
    solution = Solution()
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    result = solution.groupAnagrams(strs)

    print(result)

    print("[")
    # 既需要用到元素本身，又需要用到它的索引，直接用 enumerate() 就可以了
    # i, group -> 必须是 (索引, 元素) 的顺序
    for i, group in enumerate(result):
        # for word in group -> 'tan''nat'
        # f'"{word}"' -> '"tan"''"nat"'
        # ",".join(f'"{word}"' for word in group) -> '"tan","nat"'
        formatted_group = ",".join(f'"{word}"' for word in group)
        # 判断是否是最后一个元素，如果是最后一个元素就不加逗号，否则加逗号
        end_comma = "," if i < len(result) - 1 else ""

        print(f"  [{formatted_group}]{end_comma}")
    print("]")

